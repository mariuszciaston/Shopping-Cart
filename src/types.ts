export type ProductTypes = {
  id: number;
  image: string;
  title: string;
  price: number;
  description: string;
};

export type ProductsContextType = {
  productsData: ProductTypes[] | null;
  error: null;
  loading: boolean;
};
