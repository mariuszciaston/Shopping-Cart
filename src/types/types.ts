export type Product = {
  id: number;
  image: string;
  title: string;
  price: number;
  description: string;
};

export type ProductState = {
  productsData: Product[] | null;
  error: string | null;
  loading: boolean;
};

export type QuantitySelectorProps = {
  quantity: number;
  setQuantity: (quantity: number) => void;
};

export type CartItem = {
  id: number;
  quantity: number;
};

export type CartState = {
  itemsInCart: CartItem[];
  setItemsInCart: React.Dispatch<React.SetStateAction<CartItem[]>>;
};

export type UpdateCart = (
  id: CartItem["id"],
  quantity: CartItem["quantity"],
  itemsInCart: CartState["itemsInCart"],
  setItemsInCart: CartState["setItemsInCart"],
) => void;

export type ShoppingCart = {
  productsData: Product[] | null;
  itemsInCart: CartItem[];
};
