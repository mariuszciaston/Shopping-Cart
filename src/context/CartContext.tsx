import { CartContextType } from "@/types/types";
import { createContext, useContext } from "react";

export const CartContext = createContext<CartContextType | null>(null);

export const useCartContext = () => {
  const context = useContext(CartContext);
  if (!context) {
    throw new Error("useCartContext must be used within a CartProvider");
  }
  return context;
};
