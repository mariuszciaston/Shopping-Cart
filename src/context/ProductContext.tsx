import { ProductsContextType } from "@/types";
import { createContext, useContext } from "react";

export const ProductsContext = createContext<ProductsContextType | null>(null);

export const useProductsContext = () => {
  const context = useContext(ProductsContext);
  if (!context) {
    throw new Error(
      "useProductsContext must be used within a ProductsProvider",
    );
  }
  return context;
};
