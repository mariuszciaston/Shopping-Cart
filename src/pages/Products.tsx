import Header from "../components/Header";
import Main from "../components/Main";
import Footer from "../components/Footer";
import Wrapper from "@/components/ui/wrapper";
import { useProductsContext } from "@/context/ProductContext";
import RenderMultipleProducts from "@/components/RenderMultipleProducts";

function Products() {
  const { productsData, error, loading } = useProductsContext();

  return (
    <Wrapper>
      <Header />
      <Main>
        <RenderMultipleProducts
          productsData={productsData}
          error={error}
          loading={loading}
        />
      </Main>
      <Footer />
    </Wrapper>
  );
}

export default Products;
