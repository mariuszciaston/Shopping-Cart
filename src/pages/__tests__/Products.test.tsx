import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ProductsContext } from "@/context/ProductContext";
import Products from "@/pages/Products";
import { describe, it, expect } from "vitest";
import { Product, ProductState } from "@/types/types";
import { CartContext } from "@/context/CartContext";

const mockProductsData: Product[] = [
  {
    id: 1,
    title: "Product 1",
    price: 10,
    description: "Description for Product 1",
    image: "image1.jpg",
  },
  {
    id: 2,
    title: "Product 2",
    price: 20,
    description: "Description for Product 2",
    image: "image2.jpg",
  },
  {
    id: 3,
    title: "Product 3",
    price: 30,
    description: "Description for Product 3",
    image: "image3.jpg",
  },
  {
    id: 4,
    title: "Product 4",
    price: 40,
    description: "Description for Product 4",
    image: "image4.jpg",
  },
  {
    id: 5,
    title: "Product 5",
    price: 50,
    description: "Description for Product 5",
    image: "image5.jpg",
  },
  {
    id: 6,
    title: "Product 6",
    price: 60,
    description: "Description for Product 6",
    image: "image6.jpg",
  },
  {
    id: 7,
    title: "Product 7",
    price: 70,
    description: "Description for Product 7",
    image: "image7.jpg",
  },
  {
    id: 8,
    title: "Product 8",
    price: 80,
    description: "Description for Product 8",
    image: "image8.jpg",
  },
];

const mockCartContent = {
  itemsInCart: [
    {
      id: 1,
      quantity: 1,
    },
  ],
  setItemsInCart: () => {},
};

const renderHome = (contextValue: ProductState | null) => {
  return render(
    <ProductsContext.Provider value={contextValue}>
      <CartContext.Provider value={mockCartContent}>
        <MemoryRouter>
          <Products />
        </MemoryRouter>
      </CartContext.Provider>
    </ProductsContext.Provider>,
  );
};

describe("Products Component", () => {
  it("renders header, main, and footer components", () => {
    renderHome({ productsData: [], error: null, loading: false });
    expect(screen.getByRole("banner")).toBeInTheDocument();
    expect(screen.getByRole("main")).toBeInTheDocument();
    expect(screen.getByRole("contentinfo")).toBeInTheDocument();
  });

  it("displays loading state", () => {
    renderHome({ productsData: [], error: null, loading: true });
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("displays error message", () => {
    renderHome({
      productsData: [],
      error: "A network error was encountered",
      loading: false,
    });
    expect(
      screen.getByText("A network error was encountered"),
    ).toBeInTheDocument();
  });

  it("renders products correctly", () => {
    renderHome({ productsData: mockProductsData, error: null, loading: false });
    expect(screen.getByText("Product 1")).toBeInTheDocument();
    expect(screen.getByText("Product 2")).toBeInTheDocument();
    expect(screen.getByText("Product 3")).toBeInTheDocument();
    expect(screen.getByText("Product 4")).toBeInTheDocument();
    expect(screen.getByText("Product 5")).toBeInTheDocument();
    expect(screen.getByText("Product 6")).toBeInTheDocument();
    expect(screen.getByText("Product 7")).toBeInTheDocument();
    expect(screen.getByText("Product 8")).toBeInTheDocument();
  });

  it("matches snapshot", (contextValue: ProductState) => {
    const { asFragment } = render(
      <ProductsContext.Provider value={contextValue}>
        <CartContext.Provider value={mockCartContent}>
          <MemoryRouter>
            <Products />
          </MemoryRouter>
        </CartContext.Provider>
      </ProductsContext.Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
