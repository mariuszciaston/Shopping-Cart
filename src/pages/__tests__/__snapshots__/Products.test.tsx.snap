// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Products page component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex h-[calc(100vh)] flex-col"
    id="wrapper"
  >
    <header
      class="mx-auto flex w-full max-w-[1280px] flex-wrap items-center justify-between gap-4 p-4 sm:gap-8 sm:p-8"
    >
      <h1
        class="text-3xl font-bold"
      >
        <a
          class="rounded-md border-2 border-black"
          data-discover="true"
          href="/"
        >
          <span
            class="bg-black text-white"
          >
             fake 
          </span>
          <span>
             STORE 
          </span>
        </a>
      </h1>
      <nav
        class="order-0 sm:order-0 md:order-2"
      >
        <ul
          class="flex items-center gap-8"
        >
          <li>
            <a
              data-discover="true"
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              data-discover="true"
              href="/products"
            >
              Products
            </a>
          </li>
          <li>
            <a
              data-discover="true"
              data-testid="cart-link"
              href="/cart"
            >
              <div
                class="relative"
              >
                <svg
                  class="lucide lucide-shopping-basket h-8 w-8"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m15 11-1 9"
                  />
                  <path
                    d="m19 11-4-7"
                  />
                  <path
                    d="M2 11h20"
                  />
                  <path
                    d="m3.5 11 1.6 7.4a2 2 0 0 0 2 1.6h9.8a2 2 0 0 0 2-1.6l1.7-7.4"
                  />
                  <path
                    d="M4.5 15.5h15"
                  />
                  <path
                    d="m5 11 4-7"
                  />
                  <path
                    d="m9 11 1 9"
                  />
                </svg>
                <span
                  class="absolute -bottom-1 -right-1 flex h-5 min-w-5 items-center justify-center rounded-full bg-red-500 text-xs text-white"
                  data-testid="cart-item-count"
                >
                  1
                </span>
              </div>
            </a>
          </li>
        </ul>
      </nav>
    </header>
    <main
      class="flex-1 bg-gray-100"
    >
      <div
        class="mx-auto grid h-full w-full max-w-[1280px] content-start py-8 sm:p-8"
      >
        <section
          class="grid grid-cols-[repeat(auto-fill,minmax(240px,1fr))] content-start gap-4 sm:gap-8"
        />
      </div>
    </main>
    <footer
      class="mx-auto flex w-full max-w-[1280px] items-center justify-center p-4 text-sm sm:p-8"
    >
      <span>
        Mariusz Ciastoń 2025  
      </span>
      <a
        href="https://github.com/mariuszciaston/"
        rel="noopener"
        target="_blank"
      >
        <i
          aria-hidden="true"
          class="fab fa-github fa-2x transition-transform duration-500 ease-in-out hover:rotate-1turn hover:scale-125 hover:transform"
        />
      </a>
    </footer>
  </div>
</DocumentFragment>
`;
